packages/contracts-bedrock/src/safe/TimelockGuard.sol

```diff
@@ -0,0 +1,350 @@
+        // the transaction hash derivation.
+        bytes32 txHash = _safe.getTransactionHash(
+            _params.to,
+            _params.value,
+            _params.data,
+            _params.operation,
+            _params.safeTxGas,
+            _params.baseGas,
+            _params.gasPrice,
+            _params.gasToken,
+            _params.refundReceiver,
+            _nonce
+        );
+
+        // Check if the transaction exists
+        // A transaction can only be scheduled once, regardless of whether it has been cancelled or not.
+        if (scheduledTransactions[_safe][txHash].executionTime != 0) {
+            revert TimelockGuard_TransactionAlreadyScheduled();
+        }
+
+        // Verify signatures using the Safe's signature checking logic
+        // This function call reverts if the signatures are invalid.
+        _safe.checkSignatures(txHash, txHashData, _signatures);
+
+        // Calculate the execution time
+        uint256 executionTime = block.timestamp + safeConfigs[_safe].timelockDelay;
+
+        // Schedule the transaction
+        scheduledTransactions[_safe][txHash] =
+            ScheduledTransaction({ executionTime: executionTime, cancelled: false, executed: false });
+
+        emit TransactionScheduled(_safe, txHash, executionTime);
+    }
+
+    /// @notice Returns the list of all scheduled but not cancelled transactions for a given safe
+    /// @dev MUST NOT revert - NOT IMPLEMENTED YET
+    /// @return List of pending transaction hashes
+    function checkPendingTransactions(address) external pure returns (bytes32[] memory) {
+        return new bytes32[](0);
+    }
+
+    /// @notice Cancel a scheduled transaction if cancellation threshold is met
+    /// @dev This function aims to mimic the approach which would be used by a quorum of signers to
+    ///      cancel a partially signed transaction, which would be to sign and execute an empty
+    ///      transaction at the same nonce.
+    ///      This enables us to deterministically generate the transaction inputs for a cancellation
+    ///      transaction from the transaction being cancelled.
+    ///      In this case however we cannot use a completely empty transaction (with all inputs other than the nonce being null),
+    ///      as that would allow for the signatures used to cancel one transaction at nonce X to
+    ///      be used to cancel all transactions at nonce X.
+    ///
+    ///      Therefore we define a custom set of inputs for a cancellation transaction, based on the
+    ///      Safe's address as well as the nonce and hash of the transaction being cancelled.
+    ///
+    ///      Since the Safe's checkNSignatures function is used, the owner can use any method
+    ///      to sign the cancellation transaction inputs, including signing with a private key,
+    ///      calling the Safe's approveHash function, or EIP1271 contract signatures.
+    function cancelTransaction(
```

Comment:
Are we going to have two separate functions, one that accepts signatures as parameters, and another one that checks whether the approvals for the no-op replacement where done as transactions against the safe?

Reply:
No, that's one of the very cool thing about the Safe, and the value of relying on its internal logic. 

By submitting a signature a [prevalidated encoding](https://github.com/base/contracts/blob/7cca28be0f94134f9de1c6f409779b57d5f269b7/script/universal/Signatures.sol#L38-L43) the [Safe will know](https://github.com/safe-global/safe-smart-account/blob/186a21a74b327f17fc41217a927dea7064f74604/contracts/GnosisSafe.sol#L286-L291) that it needs to check in the `approvedHashes` mapping for the approval!